drivers/sensors/ms5611.c

```diff
@@ -0,0 +1,704 @@
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ms5611_curtime
+ *
+ * Description: Helper to get current timestamp.
+ *
+ * Return:
+ *   Timestamp in microseconds
+ ****************************************************************************/
+
+static unsigned long ms5611_curtime(void)
+{
+  struct timespec ts;
+#ifdef CONFIG_CLOCK_MONOTONIC
+  clock_gettime(CLOCK_MONOTONIC, &ts);
+#else
+  clock_gettime(CLOCK_REALTIME, &ts);
+#endif
+
+  return 1000000ull * ts.tv_sec + ts.tv_nsec / 1000;
+}
+
+/****************************************************************************
+ * Name: ms5611_sendcmd
+ *
+ * Description:
+ *   Send a command (8-bit) to MS5611
+ *
+ ****************************************************************************/
+
+static uint8_t ms5611_sendcmd(FAR struct ms5611_dev_s *priv, uint8_t cmd)
+{
+  struct i2c_msg_s msg;
+  int ret;
+
+  msg.frequency = priv->freq;
+  msg.addr      = priv->addr;
+  msg.flags     = 0;
+  msg.buffer    = &cmd;
+  msg.length    = 1;
+
+  ret = I2C_TRANSFER(priv->i2c, &msg, 1);
+  if (ret < 0)
+    {
+      snerr("I2C_TRANSFER failed: %d\n", ret);
+      return ret;
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: ms5611_read16
+ *
+ * Description:
+ *   Read 16-bit from a MS5611 register
+ *
+ ****************************************************************************/
+
+static int ms5611_read16(FAR struct ms5611_dev_s *priv, FAR uint8_t *regval)
+{
+  struct i2c_msg_s msg;
+  int ret;
+
+  msg.frequency = priv->freq;
+  msg.addr      = priv->addr;
+  msg.flags     = I2C_M_READ;
+  msg.buffer    = regval;
+  msg.length    = 2;
+
+  ret = I2C_TRANSFER(priv->i2c, &msg, 1);
+  if (ret < 0)
+    {
+      snerr("I2C_TRANSFER failed: %d\n", ret);
+      return -1;
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: ms5611_read24
+ *
+ * Description:
+ *   Read 24-bit from a MS5611 register
+ *
+ ****************************************************************************/
+
+static int ms5611_read24(FAR struct ms5611_dev_s *priv, uint8_t *regval)
+{
+  struct i2c_msg_s msg;
+  int ret;
+
+  msg.frequency = priv->freq;
+  msg.addr      = priv->addr;
+  msg.flags     = I2C_M_READ;
+  msg.buffer    = regval;
+  msg.length    = 3;
+
+  ret = I2C_TRANSFER(priv->i2c, &msg, 1);
+  if (ret < 0)
+    {
+      snerr("I2C_TRANSFER failed: %d\n", ret);
+      return -1;
+    }
+
+  return OK;
+}
+
+static inline void baro_measure_read(FAR struct ms5611_dev_s *priv,
+                                     FAR struct sensor_event_baro *baro)
+{
+  uint32_t press;
+  uint32_t temp;
+  int32_t deltat;
+  int ret;
+  uint8_t buffer[3];
+
+  /* Enforce exclusive access */
+
+  ret = nxsem_wait(&priv->exclsem);
```

Comment:
can be interrupted by a signal. should we use non-interruptable version?

Reply:
It is not an issue, because in this case the thread that is running will go on and the thread that was waiting for semaphore will just return because "ret" will be negative. BTW other drivers doesn't enforce this exclsem for measure_read(). I added it because I was thinking that fetch() and the thread worker should work at same time, but in fact this is the the right thing to do. I think in this case be interruptable by a signal will be a good side effect because you could use Ctrl+C to interrupt the sensor application easily